### DISCUSSION TITLE
pick_random() for dictionaries

### DISCUSSION THREAD

**Discussion by oeleo1** [godotengine/godot-proposals#5796] [Engine Core]
Hi,

Recently the `pick_random()` method for arrays was introduced to the core (#3454). A shortcut for the pattern `array[randi() % array.size()]` which we lived happily without for years. So now we have it an I welcome the outcome.

From experience I immediately thought about having the same for dictionaries, where the same pattern is also frequently used. It would basically be a shortcut for `dict[ dict.keys().pick_random() ]` with the new syntax, or `dict.get( dict.keys() [randi() % dict.size()] )`.

Typical use case is spawning random events, where the event id and their correcponding scenes to be instanced are in a dict of the form
```
const event_scenes : Dictionary {
  "mom" : "res://events/mama.tscn",
  "pop" : "res://events/papa.tscn",
  "sis" : "res://events/sister.tscn",
  "bro" : "res//events/brother.tscn"
}
```

A short exchange in the commit section for 3.x occurred and suggested a few things:

- Have the same `pick_random()` functionality implemented for dicts
- Rename `pick_random()` to `get_random()` as it is effectively a shortcut of a getter
- Clarify the use case for dicts, as it is not immediately obvious since dicts are key/value pairs

Hence this discussion, instead of a proposal, since I am (sorry, but I am) not going to implement any of this. I hereby suggest to get the same function for dicts with the semantics explained above: `dict.pick_random() == dict[ dict.keys() [randi() % dict.size() ] ]`.

This discussion is therefore a thumbs up for the new shortcut for arrays and a suggestion to get it (right) also for dicts.

Thanks!

**Comment by InkONat** [comment]
You should also be able to do pick_random_value()

**Reply by Spartan322** [reply]
Just use `dict.values().pick_randon()`

[Dictionary.values()](https://docs.godotengine.org/en/stable/classes/class_dictionary.html#class-dictionary-method-values)
[Array.pick_random()](https://docs.godotengine.org/en/stable/classes/class_array.html#class-array-method-pick-random)

**Comment by AThousandShips** [comment]
I'd say that this isn't really worth adding a whole new method for, as it's very easy to work around with the suggested methods, and *doesn't* offer any performance benefits as you still need to iterate over the entries, it'd only save a very little performance margin over just using `dict.keys().pick_random()`, as the internal data of the dictionary doesn't allow random access

Further I'd argue that it's a very rare thing for people to need, further supporting that this isn't needed to in core, as it can be worked around so very easily

**Comment by theXYZT** [comment]
The PR https://github.com/godotengine/godot/pull/104473 will make this possible.
